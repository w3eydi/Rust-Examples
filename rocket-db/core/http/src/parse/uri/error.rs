use std::{fmt, convert};
use std::borrow::Cow;

use pear::error::Expected;
use pear::input::ParseError;
use crate::parse::uri::RawInput;
use crate::ext::IntoOwned;

/// Error emitted on URI parse failure.
///
/// Internally, the type includes information about where the parse error
/// occurred (the error's context) and information about what went wrong.
/// Externally, this information can be retrieved (in textual form) through its
/// `Display` implementation. In other words, by printing a value of this type.
#[derive(Debug)]
pub struct Error<'a> {
    pub(crate) expected: Expected<u8, Cow<'a, [u8]>>,
    pub(crate) index: usize,
}

#[doc(hidden)]
impl<'a> From<ParseError<RawInput<'a>>> for Error<'a> {
    fn from(inner: ParseError<RawInput<'a>>) -> Self {
        let expected = inner.error.map(convert::identity, |v| v.values.into());
        Error { expected, index: inner.info.context.start }
    }
}

impl Error<'_> {
    /// Returns the byte index into the text where the error occurred if it is
    /// known.
    ///
    /// # Example
    ///
    /// ```rust
    /// # extern crate rocket;
    /// use rocket::http::uri::Origin;
    ///
    /// let err = Origin::parse("/foo bar").unwrap_err();
    /// assert_eq!(err.index(), 4);
    /// ```
    pub fn index(&self) -> usize {
        self.index
    }
}

impl fmt::Display for Error<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} at index {}", self.expected, self.index)
    }
}

impl IntoOwned for Error<'_> {
    type Owned = Error<'static>;

    fn into_owned(self) -> Error<'static> {
        Error {
            expected: self.expected.map(|t| t, |s| s.into_owned().into()),
            index: self.index
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::parse::uri::origin_from_str;

    macro_rules! check_err {
        ($url:expr => $error:expr) => {{
            let e = origin_from_str($url).unwrap_err();
            assert_eq!(e.to_string(), $error.to_string())
        }}
    }

    #[test]
    fn check_display() {
        check_err!("a" => "expected token '/' but found 'a' at index 0");
        check_err!("?" => "expected token '/' but found '?' at index 0");
        check_err!("è¿™" => "expected token '/' but found byte 232 at index 0");
    }
}
