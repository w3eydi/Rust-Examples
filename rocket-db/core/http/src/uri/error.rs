//! Errors arising from parsing invalid URIs.

use std::fmt;

pub use crate::parse::uri::Error;

/// The error type returned when a URI conversion fails.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct TryFromUriError(pub(crate) ());

impl fmt::Display for TryFromUriError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        "invalid conversion from general to specific URI variant".fmt(f)
    }
}

/// An error interpreting a segment as a [`PathBuf`] component in
/// [`Segments::to_path_buf()`].
///
/// [`PathBuf`]: std::path::PathBuf
/// [`Segments::to_path_buf()`]: crate::uri::Segments::to_path_buf()
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum PathError {
    /// The segment started with the wrapped invalid character.
    BadStart(char),
    /// The segment contained the wrapped invalid character.
    BadChar(char),
    /// The segment ended with the wrapped invalid character.
    BadEnd(char),
}
