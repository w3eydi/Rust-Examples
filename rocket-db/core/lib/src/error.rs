//! Types representing various errors that can occur in a Rocket application.

use std::{io, fmt};
use std::sync::atomic::{Ordering, AtomicBool};

use yansi::Paint;
use figment::Profile;

/// An error that occurs during launch.
///
/// An `Error` is returned by [`launch()`](crate::Rocket::launch()) when
/// launching an application fails or, more rarely, when the runtime fails after
/// lauching.
///
/// # Panics
///
/// A value of this type panics if it is dropped without first being inspected.
/// An _inspection_ occurs when any method is called. For instance, if
/// `println!("Error: {}", e)` is called, where `e: Error`, the `Display::fmt`
/// method being called by `println!` results in `e` being marked as inspected;
/// a subsequent `drop` of the value will _not_ result in a panic. The following
/// snippet illustrates this:
///
/// ```rust
/// # let _ = async {
/// if let Err(error) = rocket::build().launch().await {
///     // This println "inspects" the error.
///     println!("Launch failed! Error: {}", error);
///
///     // This call to drop (explicit here for demonstration) will do nothing.
///     drop(error);
/// }
/// # };
/// ```
///
/// When a value of this type panics, the corresponding error message is pretty
/// printed to the console. The following illustrates this:
///
/// ```rust
/// # let _ = async {
/// let error = rocket::build().launch().await;
///
/// // This call to drop (explicit here for demonstration) will result in
/// // `error` being pretty-printed to the console along with a `panic!`.
/// drop(error);
/// # };
/// ```
///
/// # Usage
///
/// An `Error` value should usually be allowed to `drop` without inspection.
/// There are at least two exceptions:
///
///   1. If you are writing a library or high-level application on-top of
///      Rocket, you likely want to inspect the value before it drops to avoid a
///      Rocket-specific `panic!`. This typically means simply printing the
///      value.
///
///   2. You want to display your own error messages.
pub struct Error {
    handled: AtomicBool,
    kind: ErrorKind
}

/// The kind error that occurred.
///
/// In almost every instance, a launch error occurs because of an I/O error;
/// this is represented by the `Io` variant. A launch error may also occur
/// because of ill-defined routes that lead to collisions or because a fairing
/// encountered an error; these are represented by the `Collision` and
/// `FailedFairing` variants, respectively.
#[derive(Debug)]
#[non_exhaustive]
pub enum ErrorKind {
    /// Binding to the provided address/port failed.
    Bind(io::Error),
    /// An I/O error occurred during launch.
    Io(io::Error),
    /// An I/O error occurred in the runtime.
    Runtime(Box<dyn std::error::Error + Send + Sync>),
    /// A valid [`Config`](crate::Config) could not be extracted from the
    /// configured figment.
    Config(figment::Error),
    /// Route collisions were detected.
    Collisions(crate::router::Collisions),
    /// Launch fairing(s) failed.
    FailedFairings(Vec<crate::fairing::Info>),
    /// Sentinels requested abort.
    SentinelAborts(Vec<crate::sentinel::Sentry>),
    /// The configuration profile is not debug but not secret key is configured.
    InsecureSecretKey(Profile),
}

impl From<ErrorKind> for Error {
    fn from(kind: ErrorKind) -> Self {
        Error::new(kind)
    }
}

impl Error {
    #[inline(always)]
    pub(crate) fn new(kind: ErrorKind) -> Error {
        Error { handled: AtomicBool::new(false), kind }
    }

    #[inline(always)]
    fn was_handled(&self) -> bool {
        self.handled.load(Ordering::Acquire)
    }

    #[inline(always)]
    fn mark_handled(&self) {
        self.handled.store(true, Ordering::Release)
    }

    /// Retrieve the `kind` of the launch error.
    ///
    /// # Example
    ///
    /// ```rust
    /// use rocket::error::ErrorKind;
    ///
    /// # let _ = async {
    /// if let Err(error) = rocket::build().launch().await {
    ///     match error.kind() {
    ///         ErrorKind::Io(e) => println!("found an i/o launch error: {}", e),
    ///         e => println!("something else happened: {}", e)
    ///     }
    /// }
    /// # };
    /// ```
    #[inline]
    pub fn kind(&self) -> &ErrorKind {
        self.mark_handled();
        &self.kind
    }
}

impl std::error::Error for Error {  }

impl fmt::Display for ErrorKind {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ErrorKind::Bind(e) => write!(f, "binding failed: {}", e),
            ErrorKind::Io(e) => write!(f, "I/O error: {}", e),
            ErrorKind::Collisions(_) => "collisions detected".fmt(f),
            ErrorKind::FailedFairings(_) => "launch fairing(s) failed".fmt(f),
            ErrorKind::Runtime(e) => write!(f, "runtime error: {}", e),
            ErrorKind::InsecureSecretKey(_) => "insecure secret key config".fmt(f),
            ErrorKind::Config(_) => "failed to extract configuration".fmt(f),
            ErrorKind::SentinelAborts(_) => "sentinel(s) aborted".fmt(f),
        }
    }
}

impl fmt::Debug for Error {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.mark_handled();
        self.kind().fmt(f)
    }
}

impl fmt::Display for Error {
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.mark_handled();
        write!(f, "{}", self.kind())
    }
}

impl Drop for Error {
    fn drop(&mut self) {
        // Don't panic if the message has been seen. Don't double-panic.
        if self.was_handled() || std::thread::panicking() {
            return
        }

        match self.kind() {
            ErrorKind::Bind(ref e) => {
                error!("Rocket failed to bind network socket to given address/port.");
                info_!("{}", e);
                panic!("aborting due to socket bind error");
            }
            ErrorKind::Io(ref e) => {
                error!("Rocket failed to launch due to an I/O error.");
                info_!("{}", e);
                panic!("aborting due to i/o error");
            }
            ErrorKind::Collisions(ref collisions) => {
                fn log_collisions<T: fmt::Display>(kind: &str, collisions: &[(T, T)]) {
                    if collisions.is_empty() { return }

                    error!("Rocket failed to launch due to the following {} collisions:", kind);
                    for &(ref a, ref b) in collisions {
                        info_!("{} {} {}", a, Paint::red("collides with").italic(), b)
                    }
                }

                log_collisions("route", &collisions.routes);
                log_collisions("catcher", &collisions.catchers);

                info_!("Note: Route collisions can usually be resolved by ranking routes.");
                panic!("routing collisions detected");
            }
            ErrorKind::FailedFairings(ref failures) => {
                error!("Rocket failed to launch due to failing fairings:");
                for fairing in failures {
                    info_!("{}", fairing.name);
                }

                panic!("aborting due to fairing failure(s)");
            }
            ErrorKind::Runtime(ref err) => {
                error!("An error occurred in the runtime:");
                info_!("{}", err);
                panic!("aborting due to runtime failure");
            }
            ErrorKind::InsecureSecretKey(profile) => {
                error!("secrets enabled in non-debug without `secret_key`");
                info_!("selected profile: {}", Paint::default(profile).bold());
                info_!("disable `secrets` feature or configure a `secret_key`");
                panic!("aborting due to insecure configuration")
            }
            ErrorKind::Config(error) => {
                crate::config::pretty_print_error(error.clone());
                panic!("aborting due to invalid configuration")
            }
            ErrorKind::SentinelAborts(ref failures) => {
                error!("Rocket failed to launch due to aborting sentinels:");
                for sentry in failures {
                    let name = Paint::default(sentry.type_name).bold();
                    let (file, line, col) = sentry.location;
                    info_!("{} ({}:{}:{})", name, file, line, col);
                }

                panic!("aborting due to sentinel-triggered abort(s)");
            }
        }
    }
}
