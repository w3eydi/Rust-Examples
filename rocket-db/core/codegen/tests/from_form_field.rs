use rocket::form::{FromFormField, ValueField, FromForm, Options, Errors};

fn parse<'v, T: FromForm<'v>>(value: &'v str) -> Result<T, Errors<'v>> {
    let mut context = T::init(Options::Lenient);
    T::push_value(&mut context, ValueField::from_value(value));
    T::finalize(context)
}

macro_rules! assert_parse {
    ($($string:expr),* => $item:ident :: $variant:ident) => ($(
        match parse::<$item>($string) {
            Ok($item::$variant) => { /* okay */ },
            Ok(item) => panic!("Failed to parse {} as {:?}. Got {:?} instead.",
                               $string, $item::$variant, item),
            Err(e) => panic!("Failed to parse {} as {}: {}",
                             $string, stringify!($item), e),

        }
    )*)
}

macro_rules! assert_no_parse {
    ($($string:expr),* => $item:ident) => ($(
        match parse::<$item>($string) {
            Err(_) => { /* okay */ },
            Ok(item) => panic!("Unexpectedly parsed {} as {:?}", $string, item)
        }
    )*)
}

#[test]
fn from_form_value_simple() {
    #[derive(Debug, FromFormField)]
    enum Foo { A, B, C, }

    assert_parse!("a", "A" => Foo::A);
    assert_parse!("b", "B" => Foo::B);
    assert_parse!("c", "C" => Foo::C);
}

#[test]
fn from_form_value_weirder() {
    #[allow(non_camel_case_types)]
    #[derive(Debug, FromFormField)]
    enum Foo { Ab_Cd, OtherA }

    assert_parse!("ab_cd", "ab_CD", "Ab_CD" => Foo::Ab_Cd);
    assert_parse!("othera", "OTHERA", "otherA", "OtherA" => Foo::OtherA);
}

#[test]
fn from_form_value_no_parse() {
    #[derive(Debug, FromFormField)]
    enum Foo { A, B, C, }

    assert_no_parse!("abc", "ab", "bc", "ca" => Foo);
    assert_no_parse!("b ", "a ", "c ", "a b" => Foo);
}

#[test]
fn from_form_value_renames() {
    #[derive(Debug, FromFormField)]
    enum Foo {
        #[field(value = "foo")]
        #[field(value = "bark")]
        Bar,
        #[field(value = ":book")]
        Book
    }

    assert_parse!("foo", "FOO", "FoO", "bark", "BARK", "BaRk" => Foo::Bar);
    assert_parse!(":book", ":BOOK", ":bOOk", ":booK" => Foo::Book);
    assert_no_parse!("book", "bar" => Foo);
}

#[test]
fn from_form_value_raw() {
    #[allow(non_camel_case_types)]
    #[derive(Debug, FromFormField)]
    enum Keyword {
        r#type,
        this,
    }

    assert_parse!("type", "tYpE" => Keyword::r#type);
    assert_parse!("this" => Keyword::this);
    assert_no_parse!("r#type" => Keyword);
}

#[test]
fn form_value_errors() {
    use rocket::form::error::{ErrorKind, Entity};

    #[derive(Debug, FromFormField)]
    enum Foo { Bar, Bob }

    let errors = parse::<Foo>("blob").unwrap_err();
    assert!(errors.iter().any(|e| {
        && "blob" == &e.value.as_ref().unwrap()
        && e.entity == Entity::Value
        && match &e.kind {
            ErrorKind::InvalidChoice { choices } => &choices[..] == &["Bar", "Bob"],
            _ => false
        }
    }));
}
