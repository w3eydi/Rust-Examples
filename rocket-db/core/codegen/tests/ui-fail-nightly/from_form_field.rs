../ui-fail/from_form_field.rs