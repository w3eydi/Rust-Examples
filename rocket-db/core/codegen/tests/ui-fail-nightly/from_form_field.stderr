error: tuple structs are not supported
 --> $DIR/from_form_field.rs:4:1
  |
4 | struct Foo1;
  | ^^^^^^^^^^^^
  |
note: error occurred while deriving `FromFormField`
 --> $DIR/from_form_field.rs:3:10
  |
3 | #[derive(FromFormField)]
  |          ^^^^^^^^^^^^^
  = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: tuple structs are not supported
 --> $DIR/from_form_field.rs:7:1
  |
7 | struct Foo2(usize);
  | ^^^^^^^^^^^^^^^^^^^
  |
note: error occurred while deriving `FromFormField`
 --> $DIR/from_form_field.rs:6:10
  |
6 | #[derive(FromFormField)]
  |          ^^^^^^^^^^^^^
  = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: named structs are not supported
  --> $DIR/from_form_field.rs:10:1
   |
10 | / struct Foo3 {
11 | |     foo: usize,
12 | | }
   | |_^
   |
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:9:10
   |
9  | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: variants cannot have fields
  --> $DIR/from_form_field.rs:16:6
   |
16 |     A(usize),
   |      ^^^^^^^
   |
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:14:10
   |
14 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: enum must have at least one variant
  --> $DIR/from_form_field.rs:20:1
   |
20 | enum Foo5 { }
   | ^^^^^^^^^^^^^
   |
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:19:10
   |
19 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: type generics are not supported
  --> $DIR/from_form_field.rs:23:11
   |
23 | enum Foo6<T> {
   |           ^
   |
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:22:10
   |
22 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid value: expected string literal
  --> $DIR/from_form_field.rs:29:21
   |
29 |     #[field(value = 123)]
   |                     ^^^
   |
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:27:10
   |
27 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: expected literal, found bare path "value"
  --> $DIR/from_form_field.rs:35:13
   |
35 |     #[field(value)]
   |             ^^^^^
   |
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:33:10
   |
33 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: variant has conflicting values
  --> $DIR/from_form_field.rs:41:5
   |
41 | /     #[field(value = "bar")]
42 | |     #[field(value = "bar")]
43 | |     A,
   | |_____^
   |
note: this value...
  --> $DIR/from_form_field.rs:41:21
   |
41 |     #[field(value = "bar")]
   |                     ^^^^^
note: ...conflicts with this value
  --> $DIR/from_form_field.rs:42:21
   |
42 |     #[field(value = "bar")]
   |                     ^^^^^
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:39:10
   |
39 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field value conflicts with previous value
  --> $DIR/from_form_field.rs:50:21
   |
50 |     #[field(value = "BAr")]
   |                     ^^^^^
   |
help: ...declared in this variant
  --> $DIR/from_form_field.rs:50:5
   |
50 | /     #[field(value = "BAr")]
51 | |     B,
   | |_____^
note: previous field with conflicting name
  --> $DIR/from_form_field.rs:48:5
   |
48 | /     #[field(value = "bar")]
49 | |     A,
   | |_____^
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:46:10
   |
46 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field value conflicts with previous value
  --> $DIR/from_form_field.rs:57:21
   |
57 |     #[field(value = "a")]
   |                     ^^^
   |
help: ...declared in this variant
  --> $DIR/from_form_field.rs:57:5
   |
57 | /     #[field(value = "a")]
58 | |     B,
   | |_____^
note: previous field with conflicting name
  --> $DIR/from_form_field.rs:56:5
   |
56 |     A,
   |     ^
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:54:10
   |
54 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: variant has conflicting values
  --> $DIR/from_form_field.rs:80:5
   |
80 | /     #[field(value = "FoO")]
81 | |     #[field(value = "foo")]
82 | |     A,
   | |_____^
   |
note: this value...
  --> $DIR/from_form_field.rs:80:21
   |
80 |     #[field(value = "FoO")]
   |                     ^^^^^
note: ...conflicts with this value
  --> $DIR/from_form_field.rs:81:21
   |
81 |     #[field(value = "foo")]
   |                     ^^^^^
note: error occurred while deriving `FromFormField`
  --> $DIR/from_form_field.rs:78:10
   |
78 | #[derive(FromFormField)]
   |          ^^^^^^^^^^^^^
   = note: this error originates in the derive macro `FromFormField` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field has conflicting names
  --> $DIR/from_form_field.rs:87:5
   |
87 | /     #[field(name = "foo")]
88 | |     #[field(name = uncased("FOO"))]
89 | |     single: usize,
   | |_________________^
   |
note: this field name...
  --> $DIR/from_form_field.rs:87:20
   |
87 |     #[field(name = "foo")]
   |                    ^^^^^
note: ...conflicts with this field name
  --> $DIR/from_form_field.rs:88:28
   |
88 |     #[field(name = uncased("FOO"))]
   |                            ^^^^^
note: error occurred while deriving `FromForm`
  --> $DIR/from_form_field.rs:85:10
   |
85 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field name conflicts with previous name
  --> $DIR/from_form_field.rs:96:20
   |
96 |     #[field(name = "foo")]
   |                    ^^^^^
   |
help: declared in this field
  --> $DIR/from_form_field.rs:96:5
   |
96 | /     #[field(name = "foo")]
97 | |     other: usize,
   | |________________^
note: previous field with conflicting name
  --> $DIR/from_form_field.rs:94:5
   |
94 | /     #[field(name = "foo")]
95 | |     single: usize,
   | |_________________^
note: error occurred while deriving `FromForm`
  --> $DIR/from_form_field.rs:92:10
   |
92 | #[derive(FromForm)]
   |          ^^^^^^^^
   = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field name conflicts with previous name
   --> $DIR/from_form_field.rs:104:5
    |
104 |     hello_there: usize,
    |     ^^^^^^^^^^^
    |
help: declared in this field
   --> $DIR/from_form_field.rs:104:5
    |
104 |     hello_there: usize,
    |     ^^^^^^^^^^^^^^^^^^
note: previous field with conflicting name
   --> $DIR/from_form_field.rs:102:5
    |
102 | /     #[field(name = uncased("HELLO_THERE"))]
103 | |     single: usize,
    | |_________________^
note: error occurred while deriving `FromForm`
   --> $DIR/from_form_field.rs:100:10
    |
100 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)

error: field name conflicts with previous name
   --> $DIR/from_form_field.rs:111:5
    |
111 |     hello_there: usize,
    |     ^^^^^^^^^^^
    |
help: declared in this field
   --> $DIR/from_form_field.rs:111:5
    |
111 |     hello_there: usize,
    |     ^^^^^^^^^^^^^^^^^^
note: previous field with conflicting name
   --> $DIR/from_form_field.rs:109:5
    |
109 | /     #[field(name = "hello_there")]
110 | |     single: usize,
    | |_________________^
note: error occurred while deriving `FromForm`
   --> $DIR/from_form_field.rs:107:10
    |
107 | #[derive(FromForm)]
    |          ^^^^^^^^
    = note: this error originates in the derive macro `FromForm` (in Nightly builds, run with -Z macro-backtrace for more info)
